### PR TITLE
Referral link

### PR DESCRIPTION
**Pull request by lamara**
Adds ability for an outside page to set ?referral=true when linking to the /join page (or the dashboard), that will display a popup asking the user if they want to go back to the original page.

### CODE REVIEW

File: components/dashboard/dashboard-controller.js [hunk 1]
```diff
@@ -11,11 +11,24 @@ define([], function() {
...
+  // Setting ?referral=true will trigger a prompt to for the user
+  // to go back to the original page that linked to this join page.
+  if($routeParams && $routeParams.referral === 'true' && document.referrer) {
+    self.state.redirectLink = document.referrer;
+    self.modals.shouldShowRedirectModal = true;
+  }
+
+  self.redirect = function() {
+    if(self.state.redirectLink) {
+      window.location = self.state.redirectLink;
```
Comment (dlongley) [inline components/dashboard/dashboard-controller.js]: Use `$window`.

File: components/dashboard/dashboard-controller.js [hunk 2]
```diff
@@ -11,11 +11,24 @@ define([], function() {
...
-function factory(brIdentityService) {
+function factory($routeParams, brIdentityService) {
   var self = this;
   self.identity = brIdentityService.identity;
   self.state = {};
   self.modals = {};
+
+  // Setting ?referral=true will trigger a prompt to for the user
+  // to go back to the original page that linked to this join page.
+  if($routeParams && $routeParams.referral === 'true' && document.referrer) {
```
Comment (dlongley) [inline components/dashboard/dashboard-controller.js]: If `$document` is a thing, use it.

File: components/dashboard/dashboard-controller.js [hunk 3]
```diff
@@ -11,11 +11,24 @@ define([], function() {
...
 
 /* @ngInject */
-function factory(brIdentityService) {
+function factory($routeParams, brIdentityService) {
   var self = this;
   self.identity = brIdentityService.identity;
   self.state = {};
   self.modals = {};
+
+  // Setting ?referral=true will trigger a prompt to for the user
```
Comment (dlongley) [inline components/dashboard/dashboard-controller.js]: typo "a prompt to for"

File: components/dashboard/dashboard.html [hunk 4]
```diff
@@ -1,5 +1,9 @@
 <div class="dashboard ng-cloak" ng-controller="DashboardController as model">
 
+  <div ng-if="model.state.redirectLink" class="row">
+    <div class="alert alert-info text-center col-md-offset-2 col-md-8">Go back to <a href="{{model.state.redirectLink}}">{{model.state.redirectLink}}</a> ?</div>
```
Comment (dlongley) [inline components/dashboard/dashboard.html]: Prepend text with: "It looks like you were referred here by another website. Go back to..."

File: components/dashboard/dashboard.html [hunk 5]
```diff
@@ -55,4 +59,17 @@ <h1 class="headline">Dashboard</h1>
     </div>
   </div>
 
+  <stackable-modal stackable="model.modals.shouldShowRedirectModal">
+    <br-modal br-title="Go back?">
+      <br-modal-body>
+        <p>We noticed that you arrived here from <a href="{{model.state.redirectLink}}">{{model.state.redirectLink}}</a></p>
+        <p>Would you like to go back?</p>
```
Comment (dlongley) [inline components/dashboard/dashboard.html]: "Would you like to go back now?" <-- added "now"

File: components/identity/create-identity-controller.js [hunk 6]
```diff
@@ -30,6 +31,10 @@ function factory(
   self.agreementAccepted = false;
   self.passphraseConfirmation = '';
 
+  // Setting ?referral=true will trigger a prompt to for the user
```
Comment (dlongley) [inline components/identity/create-identity-controller.js]: typo "a prompt to for"

File: components/idp.js [hunk 7]
```diff
@@ -86,6 +85,20 @@ module.config(function($routeProvider, routeResolverProvider) {
     })
     .when('/join', {
       title: 'Create Identity',
+      /* @ngInject */
+      resolve: {loggedIn: function($location, brSessionService, config) {
```
Comment (dlongley) [inline components/idp.js]: Move up to where resolve function is to ensure ng-annotation works properly. Change "loggedIn" to "redirectIfReferred" or something.

**Comment by lamara** [comment]
Going to add an "?auto=true" option or something similar to auto-redirect back to the referrer if someone wants to utilize that option.

**Comment by lamara** [comment]
See any issues more issues @dlongley ?

File: components/dashboard/dashboard-controller.js [hunk 8]
```diff
@@ -11,11 +11,24 @@ define([], function() {
...
+function factory($routeParams, brIdentityService, $document, $window) {
   var self = this;
   self.identity = brIdentityService.identity;
   self.state = {};
   self.modals = {};
+
+  // Setting ?referral=true will trigger a prompt for the user
+  // to go back to the original page that linked to this join page.
+  if($routeParams && $routeParams.referral === 'true' && $document.referrer) {
+    self.state.redirectLink = document.referrer;
```
Comment (dlongley) [inline components/dashboard/dashboard-controller.js]: Use `$document.referrer`

Comment (dlongley) [inline components/dashboard/dashboard-controller.js]: Remove extra `$routeParams` existence check if unnecessary.

File: components/identity/create-identity-controller.js [hunk 9]
```diff
@@ -30,6 +31,11 @@ function factory(
...
+  // to go back to the page that linked here once they are done.
+  var shouldRedirect = ($routeParams && $routeParams.referral === 'true');
+  var shouldRedirectAuto = ($routeParams && $routeParams.auto === 'true');
```
Comment (dlongley) [inline components/identity/create-identity-controller.js]: Remove `$routeParams` existence check if not necessary.

**Comment by dlongley** [comment]
Some minor comments then LGTM, pass to @mattcollier to merge once addressed.

**Comment by lamara** [comment]
Should be good now.

File: components/idp.js [hunk 10]
```diff
@@ -49,6 +48,25 @@ module.config(function($routeProvider, routeResolverProvider) {
...
+    $location, $window, $route, brSessionService, config) {
+    // return early if session is present
+    var wasReferred = ($location.search().referral === 'true');
+    var redirectAuto = ($location.search().auto === 'true');
+    return brSessionService.get()
+      .then(function(session) {
+        if(session.identity) {
+          var referred = wasReferred ? '?referral=true' : '';
+          if(redirectAuto) {
+            $window.location = document.referrer;
```
Comment (davidlehn) [inline components/idp.js]: $document

Comment (lamara) [inline components/idp.js]: $document has different semantics, just using document for a proper referral link.

**Pull request merged**